src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,315 @@
+            }
+
+            foreach (var item in value.RemovedItems)
+            {
+                if (designTimeInputs.TryGetValue(item, out var series))
+                {
+                    series?.Dispose();
+                    designTimeInputs.Remove(item);
+                    removedDesignTimeInputs.Add(item);
+                }
+                designTimeSharedInputs.Remove(item);
+            }
+
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem
+            {
+                Inputs = designTimeInputs.ToImmutable(),
+                SharedInputs = designTimeSharedInputs.ToImmutable(),
+            }, value.DataSourceVersions);
+
+            // Fire off the events
+            var buildManager = BuildManager.First().Value as VSBuildManager;
+            if (value.AddedSharedItems.Count > 0)
+            {
+                // if the shared items changed then all TempPEs are dirty, because shared items are included in all TempPEs
+                foreach (var item in designTimeInputs.Keys)
+                {
+                    buildManager.OnDesignTimeOutputDirty(item);
+                }
+            }
+            else
+            {
+                // otherwise just the ones that we've added are dirty
+                foreach (var item in addedDesignTimeInputs)
+                {
+                    buildManager.OnDesignTimeOutputDirty(item);
+                }
+            }
+            foreach (var item in removedDesignTimeInputs)
+            {
+                buildManager.OnDesignTimeOutputDeleted(item);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected override Task InitializeInnerCoreAsync(CancellationToken cancellationToken)
+        {
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem(), ImmutableDictionary.Create<NamedIdentity, IComparable>());
+
+            return Task.CompletedTask;
+        }
+
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<InputTuple>> targetBlock)
+        {
+            return ProjectDataSources.SyncLinkTo(
+                ProjectSubscriptionService.ProjectSource.SourceBlock.SyncLinkOptions(),
+                ProjectSubscriptionService.SourceItemsRuleSource.SourceBlock.SyncLinkOptions(),
+                targetBlock,
+                new DataflowLinkOptions { PropagateCompletion = true },
+                cancellationToken: ProjectAsynchronousTasksService.UnloadCancellationToken);
+        }
+
+        protected override Task<DesignTimeInputsDelta> PreprocessAsync(IProjectVersionedValue<InputTuple> input, DesignTimeInputsDelta previousOutput)
+        {
+            var project = input.Value.Item1.ProjectInstance;
+            var changes = input.Value.Item2.ProjectChanges[Compile.SchemaName];
+            
+            var addedDesignTimeInputs = ImmutableList.CreateBuilder<string>();
```

Comment:
Why are we using `ImmutableList` here and not `ImmutableArray`?

Reply:
`ImmutableArray` is superior to `IReadOnlyList` on a number of fronts so if I had choice I would always choose it for returning snapshots of data in an API.

`ImmutableList` etc. assumes that you will mutate this in some way, however `ImmutableList.CreateBuilder<string>()` is already stating "I am going to mutate this then create an immutable version" so its unclear to me why `ImmutableList` is desired here.

Reply:
> I thought Array was for when adding is rare.

It is.

Here's my mental model for `ImmutableArray<>` in case it's helpful:

- `ImmutableArray<>` is expensive to mutate (linear space/time complexity as internal array is copied)
- `ImmutableArray<>.Builder` is no more expensive to mutate than `List<>` (amortised constant space/time complexity)
- Calling `builder.ToImmutable` involves a defensive copy of the builder's internal array
- Calling `builder.MoveToImmutable` moves the builder's inner array to the returned `ImmutableArray<>` without copying, but only works if the builder is created with the correct initial size (otherwise it throws)
- `ImmutableArray<>` cannot be `null`, however `default(ImmutableArray<>)` throws when used, so you must check `IsDefault` or `IsDefaultOrEmpty` to avoid exceptions
- `ImmutableArray<>` implements `IReadOnlyList<>` but the conversion boxes
- `ImmutableArray<>`  allows allocation-free enumeration

> `ImmutableList.CreateBuilder<string>()` is already stating "I am going to mutate this then create an immutable version"

Agreed. It also says to me that the consumer of that immutable version is likely to make further changes over time. If that's not the case, `ImmutableArray<T>` is probably a better choice.

Reply:
Thank you that's awesome.

> ImmutableArray<>.Builder is no more expensive to mutate than List<> (amortised constant space/time complexity)
> Calling builder.ToImmutable involves a defensive copy of the builder's internal array

In that case it would seem like in this case (always starting with an empty builder, collection of unknown size) there would be no downside to just using a `List<>` and using `CreateRange` to convert it to an ImmutableArray at the end right?

Reply:
`CreateRange` takes `IEnumerable<>` but sniffs the fact that it can grab the length and use `CopyTo`. So in theory they should be roughly the same.

I ran a quick benchmark while waiting for a dump to copy. They allocate roughly the same amount of memory, but vary on speed.

| Method | Count | Mean | Allocated |
|------------------------------------- |------ |----------:|----------:|
| ImmutableArray_CreateRange_FromList |     3 | 182.44 ns |      76 B |
| ImmutableArray_Builder_UnknownLength |     3 | 125.56 ns |      84 B |
| ImmutableArray_Builder_KnownLength |     3 |  65.98 ns |      40 B |
| ImmutableArray_CreateRange_FromList |    30 | 714.48 ns |     444 B |
| ImmutableArray_Builder_UnknownLength |    30 | 622.25 ns |     408 B |
| ImmutableArray_Builder_KnownLength |    30 | 373.65 ns |     148 B |

Note I didn't use `MoveToImmutable` for `KnownLength`. It's still the best approach in all cases. WIth move it would probably be faster, but my file's almost finished copying...

I guess the variation in allocation is due to different initial array capacities and growth strategies, or fixed overheads. I'm not really sure. There's not much in it.

I'm surprised it's 15-30% faster to use the builder over the list. `CreateRange` does have to do a few type checks, but I wouldn't expect them to account for that much. Haven't looked further.

In short, if it's easy to use the builder, why not.